### ISSUE TITLE
`bundle remove` does not run install, with or without deprecated flag

### ISSUE THREAD

**Issue by joe-sharp** [rubygems/rubygems#5436]
### Describe the problem as clearly as you can

During my testing I have found that `bundle remove` does not change lockfiles. Not sure if it worked for a while or not, but I did find the [merged PR](https://github.com/rubygems/rubygems/pull/4889) and [closed issue](https://github.com/rubygems/rubygems/pull/4891) where the `--install` flag was deprecated. I think I have tested on most Bundler versions from 2.3.0+ but currently I am using 2.3.10 and `bundle remove` still does not automatically run install and doesn't update the lockfiles.

### Did you try upgrading rubygems & bundler?
I did just update rubygems as well.

### Post steps to reproduce the problem

Running the latest versions and testing on freshly generated Gemfile and Gemfile.lock. To make it easy for anyone to reproduce, I pulled down [thoughtbot/appraisal](https://github.com/thoughtbot/appraisal) and added `gem 'a-test-gem'` to the bottom of the Gemfile and ran `bundle install`. Then I tried removing a-test-gem by running the following:

### Which command did you run?
`bundle remove a-test-gem`

### What were you expecting to happen?
Expect lines from `Gemfile` AND `Gemfile.lock` to be removed.

### What actually happened?
Only the `Gemfile` is modified.

## Environment

```
Bundler       2.3.10
  Platforms   ruby, x86_64-darwin-21
Ruby          2.6.3p62 (2019-04-16 revision 67580) [x86_64-darwin-21]
  Full Path   $HOME/.asdf/installs/ruby/2.6.3/bin/ruby
  Config Dir  $HOME/.asdf/installs/ruby/2.6.3/etc
RubyGems      3.3.10
  Gem Home    $HOME/.asdf/installs/ruby/2.6.3/lib/ruby/gems/2.6.0
  Gem Path    $HOME/.gem/ruby/2.6.0:/Users/joesharp/.asdf/installs/ruby/2.6.3/lib/ruby/gems/2.6.0
  User Home   $HOME
  User Path   $HOME/.gem/ruby/2.6.0
  Bin Dir     $HOME/.asdf/installs/ruby/2.6.3/bin
Tools
  Git         2.32.0 (Apple Git-132)
  RVM         not installed
  rbenv       not installed
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2022-03-23
Git SHA           4bbb70e7de
Released Version  true
```

### Gemfile

```ruby
source "https://rubygems.org"

gemspec

# This here to make sure appraisal works with Rails 3.0.0.
gem "thor", "~> 0.14.0"

gem 'a-test-gem'
```

### Gemfile.lock

```
PATH
  remote: .
  specs:
    appraisal (2.4.1)
      bundler
      rake
      thor (>= 0.14.0)

GEM
  remote: https://rubygems.org/
  specs:
    a-test-gem (0.0.19)
    activesupport (6.1.5)
      concurrent-ruby (~> 1.0, >= 1.0.2)
      i18n (>= 1.6, < 2)
      minitest (>= 5.1)
      tzinfo (~> 2.0)
      zeitwerk (~> 2.3)
    concurrent-ruby (1.1.10)
    diff-lcs (1.5.0)
    i18n (1.10.0)
      concurrent-ruby (~> 1.0)
    minitest (5.15.0)
    rake (13.0.6)
    rspec (3.11.0)
      rspec-core (~> 3.11.0)
      rspec-expectations (~> 3.11.0)
      rspec-mocks (~> 3.11.0)
    rspec-core (3.11.0)
      rspec-support (~> 3.11.0)
    rspec-expectations (3.11.0)
      diff-lcs (>= 1.2.0, < 2.0)
      rspec-support (~> 3.11.0)
    rspec-mocks (3.11.0)
      diff-lcs (>= 1.2.0, < 2.0)
      rspec-support (~> 3.11.0)
    rspec-support (3.11.0)
    thor (0.14.6)
    tzinfo (2.0.4)
      concurrent-ruby (~> 1.0)
    zeitwerk (2.5.4)

PLATFORMS
  x86_64-darwin-21

DEPENDENCIES
  a-test-gem
  activesupport (>= 3.2.21)
  appraisal!
  rspec (~> 3.0)
  thor (~> 0.14.0)

BUNDLED WITH
   2.3.10
```

### appraisal.gemspec

```ruby
# -*- encoding: utf-8 -*-
$:.push File.expand_path("../lib", __FILE__)
require "appraisal/version"

Gem::Specification.new do |s|
  s.name        = 'appraisal'
  s.version     = Appraisal::VERSION.dup
  s.platform    = Gem::Platform::RUBY
  s.authors     = ['Joe Ferris', 'Prem Sichanugrist']
  s.email       = ['jferris@thoughtbot.com', 'prem@thoughtbot.com']
  s.homepage    = 'http://github.com/thoughtbot/appraisal'
  s.summary     = 'Find out what your Ruby gems are worth'
  s.description = 'Appraisal integrates with bundler and rake to test your library against different versions of dependencies in repeatable scenarios called "appraisals."'
  s.license     = 'MIT'

  s.files         = `git ls-files`.split("\n")
  s.test_files    = `git ls-files -- {test,spec,features}/*`.split("\n")
  s.executables   = `git ls-files -- bin/*`.split("\n").map{ |f| File.basename(f) }
  s.require_paths = ["lib"]

  s.required_ruby_version = ">= 2.3.0"

  s.add_runtime_dependency('rake')
  s.add_runtime_dependency('bundler')
  s.add_runtime_dependency('thor', '>= 0.14.0')

  s.add_development_dependency("activesupport", ">= 3.2.21")
  s.add_development_dependency('rspec', '~> 3.0')
end
```

**Comment by duckinator** [comment]
hey, thanks for bringing this to our attention!

I did some digging and determined:

- #4891 _does_ look like the correct fix for this.
- According to [CHANGELOG.md](https://github.com/rubygems/rubygems/blob/master/bundler/CHANGELOG.md#2228-september-23-2021), this fix was added in Bundler v2.2.28.
- It [survived into Bundler v2.3.10](https://github.com/rubygems/rubygems/blob/bundler-v2.3.10/bundler/lib/bundler/cli/remove.rb).
- master (as of a61b361) [still includes the fix](https://github.com/rubygems/rubygems/blob/a61b3611554db41022dd7f75907213ccbf2c06f9/bundler/lib/bundler/cli/remove.rb#L10-L15)

So the fix from #4891 wasn't removed, at least.

I'll dig in a bit more and see what I can find.

**Comment by duckinator** [comment]
Reproduced on Ubuntu 20.04.4 LTS with:

- RubyGems v3.1.6 + Bundler v2.3.8
- RubyGems v3.3.10 + Bundler v2.3.8
- RubyGems v3.3.10 + Bundler v2.3.10

**Comment by duckinator** [comment]
Still digging into this, still kinda baffled honestly.

I have this setup:

```
gitpod /workspace/test-5436 $ ls
Gemfile  test.gemspec
gitpod /workspace/test-5436 $ cat Gemfile 
source "https://rubygems.org"

gemspec

gem "thor", "~> 0.14.0"
gitpod /workspace/test-5436 $ cat test.gemspec 
# -*- encoding: utf-8 -*-

Gem::Specification.new do |s|
  s.name        = 'rg-issue-5436'
  s.version     = '1.0.0'
  s.authors     = ['Ellen Marie Dash']
  s.email       = ['me@duckie.co']
  s.summary     = 'awoo'
  s.license     = 'MIT'

  s.files         = ['Gemfile', 'test.gemspec']
endgitpod /workspace/test-5436 $ 
```

<details>
<summary>First, I do a <code>bundle install</code></summary>

```
gitpod /workspace/test-5436 $ ls
Gemfile  test.gemspec
gitpod /workspace/test-5436 $ bundle install
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Using rg-issue-5436 1.0.0 from source at `.`
Using thor 0.14.6
Using bundler 2.4.0.dev
Bundle complete! 2 Gemfile dependencies, 3 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
gitpod /workspace/test-5436 $ cat Gemfile.lock 
PATH
  remote: .
  specs:
    rg-issue-5436 (1.0.0)

GEM
  remote: https://rubygems.org/
  specs:
    thor (0.14.6)

PLATFORMS
  x86_64-linux

DEPENDENCIES
  rg-issue-5436!
  thor (~> 0.14.0)

BUNDLED WITH
   2.4.0.dev
gitpod /workspace/test-5436 $
```
</details>

<details>
<summary>Second, I do <code>bundle add a-test-gem</code></summary>

```
gitpod /workspace/test-5436 $ bundle add a-test-gem
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Using bundler 2.4.0.dev
Using a-test-gem 0.0.19
Using thor 0.14.6
Using rg-issue-5436 1.0.0 from source at `.`
gitpod /workspace/test-5436 $ cat Gemfile.lock 
PATH
  remote: .
  specs:
    rg-issue-5436 (1.0.0)

GEM
  remote: https://rubygems.org/
  specs:
    a-test-gem (0.0.19)
    thor (0.14.6)

PLATFORMS
  x86_64-linux

DEPENDENCIES
  a-test-gem (~> 0.0.19)
  rg-issue-5436!
  thor (~> 0.14.0)

BUNDLED WITH
   2.4.0.dev
gitpod /workspace/test-5436 $
```
</details>

So far, so good, right?

<details>
<summary>Third, I run <code>bundle remove a-test-gem</code></summary>

```
gitpod /workspace/test-5436 $ ls
Gemfile  Gemfile.lock  test.gemspec
gitpod /workspace/test-5436 $ cat Gemfile.lock 
PATH
  remote: .
  specs:
    rg-issue-5436 (1.0.0)

GEM
  remote: https://rubygems.org/
  specs:
    a-test-gem (0.0.19)
    thor (0.14.6)

PLATFORMS
  x86_64-linux

DEPENDENCIES
  a-test-gem (~> 0.0.19)
  rg-issue-5436!
  thor (~> 0.14.0)

BUNDLED WITH
   2.4.0.dev
gitpod /workspace/test-5436 $ bundle remove a-test-gem
Removing gems from /workspace/test-5436/Gemfile
a-test-gem (~> 0.0.19) was removed.
Using a-test-gem 0.0.19
Using bundler 2.4.0.dev
Using rg-issue-5436 1.0.0 from source at `.`
Using thor 0.14.6
gitpod /workspace/test-5436 $
```
</details>

But this is where things get weird. It removes it from the Gemfile. It then does what is, as far as I know, equivalent to `bundle install` &mdash; except the output explicitly states it is _using_ `a-test-gem`! Which as far as I know means it thinks that's in the Gemfile, and it's not just being left there! It almost feels like maybe the updated `Gemfile` isn't written to disk before the install runs?

**Comment by duckinator** [comment]
I changed `Bundler::CLI::Remove#run` to this:

```ruby
    def run
      raise InvalidOption, "Please specify gems to remove." if @gems.empty?

      puts 1
      puts File.read('Gemfile')
      Injector.remove(@gems, {})
      puts 2
      puts File.read('Gemfile')
      Installer.install(Bundler.root, Bundler.definition)
      puts 3
      puts File.read('Gemfile')
    end
```

But, interestingly, it confirms that it _was_ removed on disk. I'm not sure what to make of this.

```
gitpod /workspace/test-5436 $ bundle remove a-test-gem
1
source "https://rubygems.org"

gemspec

gem "thor", "~> 0.14.0"

gem "a-test-gem", "~> 0.0.19"
Removing gems from /workspace/test-5436/Gemfile
a-test-gem (~> 0.0.19) was removed.
2
source "https://rubygems.org"

gemspec

gem "thor", "~> 0.14.0"
Using a-test-gem 0.0.19
Using bundler 2.4.0.dev
Using rg-issue-5436 1.0.0 from source at `.`
Using thor 0.14.6
3
source "https://rubygems.org"

gemspec

gem "thor", "~> 0.14.0"
gitpod /workspace/test-5436 $ 
```

**Comment by duckinator** [comment]
Part of what `bundle install` runs is this:

```ruby
      definition = Bundler.definition
      definition.validate_runtime!

      installer = Installer.install(Bundler.root, definition, options)
```

which _is_ different than what `bundle remove` does, although it uses the same function:

```ruby
Installer.install(Bundler.root, Bundler.definition)
```

Unsure if this is significant.

**Comment by duckinator** [comment]
I think I determined the problem:

`Bundler::Injector.inject(...)`, used by `bundle add ...`, explicitly updates the lockfile.

`Bundler::Injector.remove(...)`, used by `bundle remove ...`, does NOT appear to update the lockfile.

In theory, I have an idea to test this &mdash; but if it works more work will need to be done another day to make a proper fix, write tests, etc.

**Comment by duckinator** [comment]
I WAS ON THE RIGHT TRACK, THOUGH! I have a fix!

Working on cleaning it up and at least starting a WIP PR; will need to add tests and such, though, and not sure I can get to that today.

**Comment by duckinator** [comment]
It works:

```
gitpod /workspace/test-5436 $ ls
Gemfile  test.gemspec
gitpod /workspace/test-5436 $ bundle install
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Using rg-issue-5436 1.0.0 from source at `.`
Using thor 0.14.6
Using bundler 2.4.0.dev
Bundle complete! 2 Gemfile dependencies, 3 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
gitpod /workspace/test-5436 $ bundle add a-test-gem
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Using bundler 2.4.0.dev
Using thor 0.14.6
Using a-test-gem 0.0.19
Using rg-issue-5436 1.0.0 from source at `.`
gitpod /workspace/test-5436 $ bundle remove a-test-gem
Removing gems from /workspace/test-5436/Gemfile
a-test-gem (~> 0.0.19) was removed.
Using bundler 2.4.0.dev
Using rg-issue-5436 1.0.0 from source at `.`
Using thor 0.14.6
gitpod /workspace/test-5436 $
```

**Comment by duckinator** [comment]
@joe-sharp this should be resolved by #5443. That PR still needs tests, so it won't be ready just yet, however.

In the meantime, a workaround is to run `bundle install` after you use `bundle remove ...`.
